Source/Core/Core/State.cpp

```diff
@@ -554,7 +629,99 @@ u64 GetUnixTimeOfSlot(int slot)
-  return static_cast<u64>(header.time * MS_PER_SEC) + (DOUBLE_TIME_OFFSET * MS_PER_SEC);
+  return static_cast<u64>(header.legacy_header.time * MS_PER_SEC) +
+         (DOUBLE_TIME_OFFSET * MS_PER_SEC);
+}
+
+static void DecompressLZ4(std::vector<u8>& raw_buffer, u64 size, File::IOFile& f)
+{
+  raw_buffer.resize(size);
+
+  u64 total_bytes_read = 0;
+  while (true)
+  {
+    s32 compressed_data_len;
+    f.ReadArray(&compressed_data_len, 1);
+
+    auto compressed_data = std::make_unique<char[]>(compressed_data_len);
+    f.ReadBytes(compressed_data.get(), compressed_data_len);
+
+    // We need to specify the output buffer's size for safety. This may exceed the positive bound
+    // of int, causing the buffer size to be interpreted as a negative value.
+    u32 max_decompress_size = (u32)std::min((u64)LZ4_MAX_INPUT_SIZE, size - total_bytes_read);
+
+    int bytes_read =
+        LZ4_decompress_safe(compressed_data.get(), (char*)raw_buffer.data() + total_bytes_read,
+                            compressed_data_len, max_decompress_size);
+
+    if (bytes_read < 0)
+    {
+      PanicAlertFmtT("Internal LZ4 Error - decompression failed ({0}, {1}, {2}) \n"
+                     "Try loading the state again",
+                     bytes_read, compressed_data_len, max_decompress_size);
+      break;
+    }
+
+    total_bytes_read += bytes_read;
+
+    if (total_bytes_read == size)
+    {
+      break;
+    }
+    else if (total_bytes_read > size)
+    {
+      PanicAlertFmtT("Internal LZ4 Error - payload size mismatch ({0} / {1})) \n"
+                     "Try loading the state again",
+                     total_bytes_read, size);
+    }
+  }
+}
+
+static bool ValidateHeaders(const StateHeader& header)
+{
+  bool success = true;
+
+  // Game ID
+  if (strncmp(SConfig::GetInstance().GetGameID().c_str(), header.legacy_header.game_id, 6))
+  {
+    Core::DisplayMessage(fmt::format("State belongs to a different game (ID {})",
+                                     std::string_view{header.legacy_header.game_id,
+                                                      std::size(header.legacy_header.game_id)}),
+                         2000);
+    return false;
+  }
+
+  // Check both the state version and the revision string
+  std::string current_str = Common::GetScmRevStr();
+  std::string loaded_str = header.version_string;
+  const u32 loaded_version = header.version_header.version_cookie - COOKIE_BASE;
+
+  if (s_old_versions.count(loaded_version))
+  {
+    // This is a REALLY old version, before we started writing the version string to file
+    success = false;
+
+    std::pair<std::string, std::string> version_range = s_old_versions.find(loaded_version)->second;
+    std::string oldest_version = version_range.first;
+    std::string newest_version = version_range.second;
+
+    loaded_str = "Dolphin " + oldest_version + " - " + newest_version;
+  }
+  else if (current_str != loaded_str || loaded_version != STATE_VERSION)
```

Comment:
Err... I think this `current_str != loaded_str` breaks loading savestates from a different build even when the STATE_VERSION is the same, which should not be the case.

Reply:
Oh my mistake, for some reason I thought it was intended behavior that we disallow savestates across Dolphin revisions, even if the state version is the same. Yeah I need to remove this check.